src/NuGet.Clients/PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs

```diff
@@ -0,0 +1,135 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using NuGet.Common;
+using NuGet.Credentials;
+using NuGet.PackageManagement.UI;
+using NuGet.ProjectManagement;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    [Export(typeof(ICredentialServiceProvider))]
+    public class DefaultVSCredentialServiceProvider : ICredentialServiceProvider
+    {
+
+        private readonly Lazy<INuGetUILogger> _outputConsoleLogger;
+        private readonly IServiceProvider _serviceProvider;
+
+        [ImportingConstructor]
+        internal DefaultVSCredentialServiceProvider(
+           [Import(typeof(SVsServiceProvider))]
+            IServiceProvider serviceProvider,
+           Lazy<INuGetUILogger> outputConsoleLogger
+            )
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+            if (outputConsoleLogger == null)
+            {
+                throw new ArgumentNullException(nameof(outputConsoleLogger));
+            }
+            _serviceProvider = serviceProvider;
+            _outputConsoleLogger = outputConsoleLogger;
+        }
+
+        public NuGet.Configuration.ICredentialService GetCredentialService()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
```

Comment:
It was when it was running on package load. This method engages service provider and DTE very aggressively. This may cause a deadlock if not invoked on UI thread.